### PR TITLE
Merged the redux update and updated react-router, react-redux-router …

### PR DESCRIPTION
**Pull request by cramhead**
Merged the redux update and updated react-router, react-redux-router as well as devtools

Connected to rangle/rangle-starter#39
Connected to rangle/rangle-starter#48

### CODE REVIEW

File: src/store/configure-store.ts [hunk 1]
```diff
@@ -11,47 +11,60 @@ import rootReducer from '../reducers';
 // webpack-hot-loader sets some extra attributes on node's `module` if that
 // module has been hot-loaded in the browser.
 interface HotNodeModule extends NodeModule {
- hot: { accept: Function };
+    hot: { accept: Function };
 };
 
 // This global is used to turn on redux dev tools when in dev mode.
 declare let __DEV__: boolean;
+declare let __DEVTOOLS__: boolean;
```
Comment (cramhead) [inline src/store/configure-store.ts]: Whoops. This should be removed. It was there, linked from the webpack.config to allow the user to enable the DevTools via the configuration. It didn't work as it needed to conditionally add the DevTool component too

File: src/store/configure-store.ts [hunk 2]
```diff
@@ -11,47 +11,60 @@ import rootReducer from '../reducers';
...
-    });
-  }
-
-  // Required for replaying actions from devtools to work
-  if (__DEV__) {
-    reduxRouterMiddleware.listenForReplays(store);
-  }
-
-  return store;
+    let middleware = applyMiddleware(reduxRouterMiddleware, promiseMiddleware, thunk, logger);
```
Comment (SethDavenport) [inline src/store/configure-store.ts]: This will enable `logger` in prod as well.

File: tslint.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     "indent": [true, "spaces"],
     "label-position": true,
     "label-undefined": true,
-    "max-line-length": [true, 80],
+    "max-line-length": [true, 120],
```
Comment (SethDavenport) [inline tslint.json]: I've been resisting this for a while - I like my side-by-side diffs :)

File: tslint.json [hunk 4]
```diff
@@ -53,6 +54,6 @@
       "check-separator",
       "check-type"
     ],
-    "quotemark": [ true, "single", "jsx-double" ]
+    "quotemark": [ true, "single", "jsx-double", "avoid-escape" ]
```
Comment (SethDavenport) [inline tslint.json]: Not sure how I feel about this - I'm fine with escapes, typically.

**Comment by SethDavenport** [comment]
Thanks for taking the time to do this.  I'd like to get @bertrandk 's take on it too since he's more familiar with the devtools than I am.

**Comment by SethDavenport** [comment]
By the way - one of my colleagues brought this to my attention - https://github.com/zalmoxisus/redux-devtools-extension

Haven't had a chance to play with it yet but it looks like it could be a lower-impact solution.

File: src/store/configure-store.ts [hunk 5]
```diff
@@ -11,47 +11,60 @@ import rootReducer from '../reducers';
...
+
+    if (__DEV__) {
+        let devTools;
+        /* tslint:disable:no-string-literal */
+        if (window['devToolsExtension']) {
```
Comment (cramhead) [inline src/store/configure-store.ts]: If the chrome extension is installed it is automatically initialized.

File: src/store/configure-store.ts [hunk 6]
```diff
@@ -11,47 +11,60 @@ import rootReducer from '../reducers';
...
+            devTools = window['devToolsExtension']();
+        }
+        /* tslint:enable:no-string-literal */
+
+        /**
+         * if you want integrated devtools add <DevTools /> in the ../containers/app.tsx just before the closing div 
+         * and uncomment the following line
+         * devTools = require('../containers/dev-tools').default.instrument();
+         */
+        devTools = require('../containers/dev-tools').default.instrument();
```
Comment (cramhead) [inline src/store/configure-store.ts]: This line initializes the web-based devtools. The comment above it shows how users can use it. I adjusted the readme with a little explanation.

**Comment by cramhead** [comment]
I'm pretty new to the whole react-redux ecosystem so feedback would be
great for me too

File: src/store/configure-store.ts (hunk 2 above)
Comment (cramhead) [inline src/store/configure-store.ts]: Ah, ok. I wasn't sure if the logger was to be used in production. It's just
for debugging redux then?

Marc

File: tslint.json (hunk 3 above)
Comment (kosz) [inline tslint.json]: yea this makes it uncomfortable to have 3 files open side by side :-1:

File: src/store/configure-store.ts [hunk 7]
```diff
@@ -11,47 +11,60 @@ import rootReducer from '../reducers';
...
 const storageConfig = {
-  key: 'typescript-react-redux-seed',
-  serialize: (store) => {
-    return store && store.session ?
-      JSON.stringify(store.session.toJS()) : store;
-  },
-  deserialize: (state) => ({
-    session: state ? fromJS(JSON.parse(state)) : fromJS({}),
-  }),
+    key: 'typescript-react-redux-seed',
```
Comment (kosz) [inline src/store/configure-store.ts]: I see some 4 space indent here. http://rangle.io/guidelines/#2.1.1

File: src/store/configure-store.ts [hunk 8]
```diff
@@ -11,47 +11,60 @@ import rootReducer from '../reducers';
...
-
-  // Required for replaying actions from devtools to work
-  if (__DEV__) {
-    reduxRouterMiddleware.listenForReplays(store);
-  }
-
-  return store;
+    let middleware = applyMiddleware(reduxRouterMiddleware, promiseMiddleware, thunk, logger);
+
+    if (__DEV__) {
```
Comment (kosz) [inline src/store/configure-store.ts]: more 4 space ident, can we have 2 for consistency ? http://rangle.io/guidelines/#2.1.1

**Comment by SethDavenport** [comment]
Hi @cramhead,

I finally found the time to play with the chrome extension version of redux devtools and I like it.  It's much less intrusive to the codebase.  I would like both starters to go ahead and use the extension version only for now.

I've enabled it in the Babel version of the react starter: rangle/react-redux-starter/pull/89.

As for this PR, let's do the same thing.  Specifically:

- Rebase and resolve merge conflicts
- Get rid of any non-extension devtool stuff
- Make configure-store look like this: https://github.com/SethDavenport/react-redux-starter/blob/feat/extension-dev-tools/src/store/configure-store.js (with allowances for TypeScript)
- Fix the four-space indents @kosz pointed out.

Make sense?  If this is too much effort let me know and I'll take care of it.

**Comment by cramhead** [comment]
Hi @SethDavenport,
The last three things I understand what and how to do, but "Rebase and resolve merge conflicts" I'm not sure how to do. Would you have a reference I could use?

**Comment by SethDavenport** [comment]
Sure.

I see you're working off a fork.  Do you have a remote set up called 'upstream'?

```
git remote -v
```

When I do this I get:

```
origin	git@github.com:SethDavenport/typescript-react-redux-starter.git (fetch)
origin	git@github.com:SethDavenport/typescript-react-redux-starter.git (push)
upstream	git@github.com:rangle/typescript-react-redux-starter.git (fetch)
upstream	git@github.com:rangle/typescript-react-redux-starter.git (push)
```

Here origin is my fork and upstream is the original repository.

If you don't have upstream pointing to the original repo, you can set it up like this:

```sh
git remote add upstream git@github.com:rangle/typescript-react-redux-starter.git
```

Then to update your feature branch with recent changes to the main repo, do:

```sh
git checkout update-router-devtools
git fetch upstream
git rebase upstream/master
```

This will get all the recent commits from upstream master, put them on your fork's feature branch, and replay your commit on top.

You'll have some merge conflicts - not 100% sure where but probably mostly in `package.json` because greenkeeper has updated a bunch of packages recently.

Open up your editor and deal with the conflicts. Then, `git add` the conflicted files and finish the rebase with

`git rebase --continue`

At this point you should be good to go.

If the rebase gets complicated and you want to back out, just to `git rebase --abort` to return to where you started from.

**Comment by SethDavenport** [comment]
If you run into TypeScript fun with `__DEV__` and `window.devToolsExtension`, you can take a look at what I did here: https://github.com/SethDavenport/angular2-redux-starter/blob/feat/extension-dev-tools/src/store/dev-types.d.ts

**Comment by JaKXz** [comment]
I also suggest installing [EditorConfig for your editor](http://editorconfig.org/#download), it keeps the whitespace in check :)

**Comment by SethDavenport** [comment]
Closing in favour of #60 and #61 .

**Pull request closed without merging**